webidls/Browser.webidl

```diff
@@ -0,0 +1,115 @@
+/**
+  * Browser is the equivalent of a top level Constellation::Frame.
+  * Browser is a top level browsing context.
+  * ~= Tab, ~= Webview
+  */
+
+[Constructor(DOMString contextName, StorageSession storageSession)]
+interface Browser {
+  // FIXME: how is this attached to a viewport? For swapping, it's important to be able to attach and detach
+
+  Promise<void> setHandler(BrowserHandler handler);
+
+  // Used to expose special JS APIs.
+  // For example, can be used to expose the Browser API or Web Extensions APIs for content
+  // See: https://developer.chrome.com/extensions/content_scripts#execution-environment
+  Promise<void> addJSModuleResolver(/*FIXME*/, boolean onlyForFrameScript);
+
+  void destroy();
+
+  // Can't change after browser creation
+  readonly attribute DOMString browsingContextName;
+  readonly attribute StorageSession storageSession;
+
+  // Will fail if entries already exist.
+  // This can be used for session restore, or to undo tab-close.
+  // On success, historyEntries is filled, active entry's pipeline is NOT pending.
+  // If necessary, other pipelines can be restored via HistoryEntry.restorePipeline()
+  Promise<void> restoreEntries(FrozenArray<LoadData> data, unsigned long activeIndex);
+
+  readonly attribute Sequence<HistoryEntry> historyEntries;
+  readonly attribute unsigned long activeEntryIndex;
+  Promise<void> setActiveEntryIndex(unsigned long);
+
+  // Use to load a new URL. will create a new pipeline and navigate to the
+  // pipeline once not pending
+  Promise<HistoryEntry> navigate(LoadData loadData, optional Pipeline opener);
```

Comment:
When you click on a link, the switch to the new link is not happening right away, right? And during that time, the history doesn't actually change. The document is being creating, and we are waiting for HTTP metadata. The page/pipeline is "pending".

`createPrerenderingPipeline` create a special type of pipeline. One that we might never navigate to. Such pipeline has plenty of constrains, like no script being executed, limited features (no video), not XHR, …

It is used for example while you type a URL in the urlbar. Before the user press enter, we can create such a pipeline to prerender it. One the user press enter, we navigate to that pipeline.

Because the user might never navigate to this page, it has to be a special kind of page (the constrains I'm talk about).